util/timer.go

```diff
@@ -4,10 +4,33 @@ import (
 	"time"
 )
 
-// ResetTimer resets the given timer properly
-func ResetTimer(timer *time.Timer, duration time.Duration) {
-	if !timer.Stop() {
-		<-timer.C
+type SafeTimer struct {
+	timer   *time.Timer
+	timeout time.Duration
+}
+
+func NewSafeTimer(timeout time.Duration) *SafeTimer {
+	return &SafeTimer{
+		timer:   time.NewTimer(timeout),
+		timeout: timeout,
 	}
-	timer.Reset(duration)
+}
+
+func (stimer *SafeTimer) C() <-chan time.Time {
+	return stimer.timer.C
+}
+
+func (stimer *SafeTimer) Reset() {
+	// Correct timer stopping from https://github.com/golang/go/issues/27169
+	if !stimer.timer.Stop() {
+		select {
+		case <-stimer.timer.C:
+		default:
+		}
+	}
+	stimer.timer.Reset(stimer.timeout)
+}
+
+func (stimer *SafeTimer) Stop() {
```

Comment:
Aren't you supposed to dran the channel here, on all stops, rather than only on resets though?

Reply:
updated with tests